example-plugins/inj-tyfam-plugin/src/Plugin/InjTyFam.hs

```diff
@@ -0,0 +1,242 @@
+-- discharge that spurious constraint, so this doesn't lead to
+-- divergence.
+
+simplifyG :: FskEnv -> Ct -> M R
+simplifyG fskEnv ct =
+    case predTree of
+      Type.ClassPred{}  -> skip   -- TODO don't need to unpack SCs, right?
+      Type.IrredPred{}  -> skip   -- TODO this is unreachable, right?
+      Type.ForAllPred{} -> skip   -- TODO this is unreachable, right?
+      Type.EqPred eqRel lty rty -> case eqRel of
+          Type.ReprEq -> skip   -- TODO anything useful to do here?
+          Type.NomEq  -> case splitter fskEnv lty rty of
+              Nothing         -> skip   -- not the constraint we're looking for
+              Just (tc, args) -> do
+                  mbChanges <- go [] [] args
+                  case mbChanges of
+                    Nothing -> pure $ TcRnTypes.TcPluginContradiction [ct]
+                    Just x  -> ok lty tc x
+  where
+    predTree :: Type.PredTree
+    predTree = Type.classifyPredType (TcRnTypes.ctPred ct)
+
+    -- 'Nothing' means contradiction
+    -- 'Just' means (updated rtys, new constraints)
+    go :: [Xi] -> [Ct] -> [Triple] -> M (Maybe ([Xi], [Ct]))
+    go acc1 acc2 = \case
+        MkTriple lty rty i : args
+
+            -- nothing to learn at this argument position
+            | not i || Type.eqType lty rty ->
+            go (rty : acc1) acc2 args
+
+            -- contradiction!
+            | Unify.typesCantMatch [(lty, rty)] ->
+            pure Nothing
+
+            -- emit new constraint and update the RHS argument
+            | otherwise -> do
+            new <- impliedGivenEq ct lty rty
+            go (lty : acc1) (new : acc2) args
+
+        [] -> pure $ Just (reverse acc1, acc2)
+
+    -- replace the equality with one with updated RHS arguments
+    ok :: TcType -> TyCon -> ([Xi], [Ct]) -> M R
+    ok lty tc (rtys', news)
+        | null news = skip   -- we made no changes
+        | otherwise = do
+        new <- impliedGivenEq ct lty rty'
+        pure $ TcRnTypes.TcPluginOk [(ev, ct)] (new:news)
+      where
+        rty' :: TcType
+        rty' = Type.mkTyConApp tc rtys'
+
+        ev :: EvTerm
+        ev = TcRnTypes.ctEvTerm $ TcRnTypes.cc_ev ct
+
+-- | Emit a new Given equality constraint implied by another Given
+-- equality constraint
+
+impliedGivenEq :: Ct -> TcType -> TcType -> M Ct
+impliedGivenEq ct lty rty = do
+    let
+      new_co :: Coercion
+      new_co = Coercion.mkUnivCo
+        (TyCoRep.PluginProv "inj-tyfam-plugin") TyCon.Nominal lty rty
+
+    -- TODO how to incorporate @ctEvId ct@? (see #15248 on GitLab ghc)
+    new_ev <- TcPluginM.newGiven
+      (TcRnTypes.bumpCtLocDepth (TcRnTypes.ctLoc ct))  -- TODO don't bump?
+      (Type.mkPrimEqPredRole TyCon.Nominal lty rty)
+      (CoreSyn.Coercion new_co)
+
+    pure $ TcRnTypes.mkNonCanonical new_ev
+
+-----
+
+-- | See 'mkFskEnv'
+
+type FskEnv = VarEnv (TyCon, [Xi])
+
+-- | An map created from the @FunEq@s
+--
+-- NOTE not necessarily idempotent
```

Comment:
In which sense? `mkFskEnv (mkFskEnv ...)` does not type-check, so the question of whether it could be idempotent does not make sense, and similarly, I cannot use the resulting `[Xi]` to lookup a `TyCon` in the same map a second type.

Reply:
My mistake, I misread this as `data FskEnv = VarEnv [(TyCon, [Xi])]`.